sys/ztimer/core.c

```diff
@@ -406,3 +406,31 @@ static void _ztimer_print(const ztimer_clock_t *clock)
     } while ((entry = entry->next));
     puts("");
 }
+
+uint32_t ztimer_until(ztimer_clock_t *clock, ztimer_t *timer)
+{
+    unsigned state = irq_disable();
+
+    if (!_is_set(clock, timer)) {
+        irq_restore(state);
+        return 0;
+    }
+    uint32_t sum = clock->list.offset;
+    ztimer_base_t *i = clock->list.next;
+    ztimer_base_t *timer_base = &timer->base;
+
+    while (i && i != timer_base) {
+        sum += i->offset;
+        i = i->next;
+    }
+    if (i == timer_base) {
+        sum += i->offset;
+        irq_restore(state);
+        uint32_t now = (uint32_t)ztimer_now(clock);
+        return (sum > now) ? (sum - now) : 0;
```

Comment:
Both sum and now are absolute points in time whose sequence is unknown, I don't think it's meaningful to compare them.

Say the timer was set at 2\**32-100 for 95 ticks (to trigger at 2\**32-5), but we've been spending the time from 2\**32-10 to 2\**32+10 in ISRs (which is a reasonable time to spend). Still in the ISR, sum is 2\**32-5, now is 10, and we return 2\**32-15 rather than 0.

I think what would be correct to do would be to start sum at 0 (which is guaranteed not to wrap by construction of the timer chain, because nothing goes in more than 2\**32 in the future), then calculate `elapsed = now() - clock->list.offset` (which is guaranteed to not wrap by the requirement that ztimer ISRs need a chance to fire at least twice a cycle), and *then* we can `return (sum >= elapsed) ? sum - elapsed : 0;`.